src/System.Device.Gpio/System/Device/ComponentInformation.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+#pragma warning disable CS1591
+namespace System.Device
+{
+    public record ComponentInformation
+    {
+        private readonly List<ComponentInformation> _subComponents;
+
+        public ComponentInformation(object instance, string name)
+            : this(instance, name, ComponentState.Active)
+        {
+        }
+
+        public ComponentInformation(object instance, string name, ComponentState componentState)
+        {
+            if (instance == null)
+            {
+                throw new ArgumentNullException(nameof(instance));
+            }
+
+            Instance = instance;
+
+            Type = instance.GetType().FullName!;
+            Name = name;
+
+            State = componentState;
+
+            _subComponents = new List<ComponentInformation>();
+            Properties = new Dictionary<string, string>();
+        }
+
+        public string Type
+        {
+            get;
+        }
+
+        protected object Instance
+        {
+            get;
+        }
+
+        public string Name
+        {
+            get; init;
+        }
+
+        public ComponentState State
+        {
+            get; init;
+        }
+
+        public IDictionary<string, string> Properties
+        {
+            get;
+            init;
+        }
+
+        public IReadOnlyList<ComponentInformation> SubComponents => _subComponents.AsReadOnly();
+
+        public void AddSubComponent(ComponentInformation subComponent)
+        {
+            if (_subComponents == null)
```

Comment:
If the level of the subcomponents is greater than X, abort (circular dependencies). Or check in printing!